File: Frontend/lib/features/auth/domain/entities/login_response.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'login_response.freezed.dart';
part 'login_response.g.dart';

/// Modelo de respuesta del backend tras un login exitoso.
/// Mapea el DTO LoginResponse del backend Spring Boot.
///
/// El cliente guarda [token] en SecureStorage y [roles] para decidir
/// qué pantalla inicial mostrar (admin → dashboard, teacher → horario, etc.)
@freezed
class LoginResponse with _$LoginResponse {
  const LoginResponse._(); // Constructor privado para permitir métodos personalizados

  const factory LoginResponse({
    /// ID único del usuario en la base de datos (clave primaria).
    required int id,

    /// Token JWT firmado con HS512 para incluir en peticiones futuras.
    required String token,

    /// Tipo de token, siempre "Bearer".
    required String tipo,

    /// Username del usuario autenticado.
    required String username,

    /// Nombre completo: "Isabel Fernández Ruiz"
    required String nombre,

    /// Roles de Spring Security: ["ROLE_ADMIN"], ["ROLE_TEACHER"], ["ROLE_STUDENT"]
    required List<String> roles,

    /// Emoji o URL de avatar para la cabecera de navegación.
    required String avatar,

    /// Preferencia de tema: "light" o "dark".
    @Default('light') String theme,

    /// IDs de los cursos de los que este profesor es tutor.
    /// Vacío para alumnos y admins.
    @Default([]) List<int> cursosTutorIds,
  }) = _LoginResponse;

  factory LoginResponse.fromJson(Map<String, dynamic> json) =>
      _$LoginResponseFromJson(json);

  /// Devuelve el rol principal (el primero de la lista).
  String get rolPrincipal => roles.isNotEmpty ? roles.first : '';

  /// True si el usuario es administrador.
  bool get esAdmin   => roles.contains('ROLE_ADMIN');

  /// True si el usuario es profesor (docente).
  bool get esDocente => roles.contains('ROLE_TEACHER');

  /// Alias compatible con el router
  bool get esProfesor => esDocente;

  /// True si el usuario es alumno.
  bool get esAlumno  => roles.contains('ROLE_STUDENT');
}

File: Backend/src/main/java/com/gaula/dto/auth/LoginResponse.java
package com.gaula.dto.auth;

import java.util.List;

/**
 * DTO de respuesta al login exitoso.
 * Devuelto en POST /api/auth/login con status 200 OK:
 * {
 *   "token": "eyJhbGci...",
 *   "tipo": "Bearer",
 *   "username": "ifernandez",
 *   "nombre": "Isabel Fernández Ruiz",
 *   "roles": ["ROLE_TEACHER"],
 *   "avatar": "👩‍🏫"
 * }
 *
 * El cliente Flutter almacena `token` en flutter_secure_storage
 * y lo añade a cada request como "Authorization: Bearer <token>".
 */
public record LoginResponse(
    /** ID único del usuario en la base de datos (clave primaria). */
    long id,

    /** Token JWT para incluir en peticiones futuras. */
    String token,

    /** Tipo de token (siempre "Bearer"). */
    String tipo,

    /** Username del usuario autenticado. */
    String username,

    /** Nombre completo para mostrar en la UI. */
    String nombre,

    /**
     * Roles del usuario (ej: ["ROLE_ADMIN"], ["ROLE_TEACHER"], ["ROLE_STUDENT"]).
     * El cliente Flutter usa este campo para mostrar la navegación correcta.
     */
    List<String> roles,

    /** Emoji o URL de avatar para la cabecera de la app. */
    String avatar,

    /** Preferencia de tema: "light" o "dark". */
    String theme,

    /**
     * IDs de los cursos de los que este profesor es tutor.
     * Null o vacío para alumnos y admins.
     * Usado en Flutter para mostrar/ocultar acciones de tutor.
     */
    List<Long> cursosTutorIds
) {
    /** Constructor conveniente con tipo "Bearer" por defecto. */
    public static LoginResponse of(long id, String token, String username, String nombre,
                                   List<String> roles, String avatar, String theme) {
        return new LoginResponse(id, token, "Bearer", username, nombre, roles, avatar, theme, List.of());
    }

    /** Constructor con cursosTutorIds para profesores. */
    public static LoginResponse ofProfesor(long id, String token, String username, String nombre,
                                           List<String> roles, String avatar, String theme,
                                           List<Long> cursosTutorIds) {
        return new LoginResponse(id, token, "Bearer", username, nombre, roles, avatar, theme,
                                 cursosTutorIds != null ? cursosTutorIds : List.of());
    }
}
